### ISSUE TITLE
HandleNestTypesOnSourceJsonConverter does not serialize/deserialize the join field as expected (6.3.0)

### ISSUE THREAD

**Issue by badpp** [elastic/elasticsearch-net#3403]
**NEST/Elasticsearch.Net version**: 6.3.0

Already reported on version 6.2.0, see https://github.com/elastic/elasticsearch-net/issues/3356

**Elasticsearch version**: 6.x

**Description of the problem including expected versus actual behavior**:

HandleNestTypesOnSourceJsonConverter does not serialize/deserialize the join field as expected.

```
"myparent" vs "\"myparent\""
```

**Steps to reproduce**:
 1. Check attached source code
```
            var parentDocument = new MyParent
            {
                Id = 1,
                ParentProperty = "a parent prop",
                MyJoinField = "myparent"
            };

            var settings = new JsonSerializerSettings();
            settings.Converters.Add(new HandleNestTypesOnSourceJsonConverter(((IConnectionSettingsValues)connectionSettings).RequestResponseSerializer));

            var json01 = Newtonsoft.Json.JsonConvert.SerializeObject(parentDocument, Formatting.Indented, settings);
            var back01 = Newtonsoft.Json.JsonConvert.DeserializeObject<MyParent>(json01, settings);

            var json02 = Newtonsoft.Json.JsonConvert.SerializeObject(back01, Formatting.Indented, settings);

            Console.WriteLine($"JSON01: {json01}");
            Console.WriteLine($"JSON02: {json02}");
```

 2. Result

```

    JSON01: {
        "ParentProperty": "a parent prop",
        "Id": 1,
        "MyJoinField": "myparent"
    }
    JSON02: {
        "ParentProperty": "a parent prop",
        "Id": 1,
        "MyJoinField": "\"myparent\""
    }
```
 
Note: This is causing a lot of troubles for as as we use to serialize/deserialize to send documents between micro services.

**Comment by badpp** [comment]
[Program.txt](https://github.com/elastic/elasticsearch-net/files/2380089/Program.txt)

**Comment by badpp** [comment]
The problem is also showing up when using a customized source serializer (based on ConnectionSettingsAwareSerializerBase) and getting the document with a search request.

1. See attached source ProgrammCustomSourceSerializer.txt.

[ProgrammCustomerSourceSerializer.txt](https://github.com/elastic/elasticsearch-net/files/2380850/ProgrammCustomerSourceSerializer.txt)

2. Result
`
JSON01: {
  "ParentProperty": "a parent prop",
  "Id": 2,
  "JoinField": "myparent"
}
JSON02: {
  "ParentProperty": "a parent prop",
  "Id": 2,
  "JoinField": "\"myparent\""
}
`
3. Using the document from the search for an following index operation will fail with
`
            var indexResult = client.Index(searchResult.Documents.First(), s => s.Index(index).Type("mydocument").Refresh(Refresh.True));
            Console.WriteLine(indexResult.DebugInformation);
`
`
{"error":{"root_cause":[{"type":"mapper_parsing_exception","reason":"failed to parse"}],"type":"mapper_parsing_exception","reason":"failed to parse","caused_by":{"type":"illegal_argument_exception","reason":"unknown join name [\"myparent\"] for field [joinField]"}},"status":400}
`

**Comment by baddonkey** [comment]
I assume this bug was fixed on the master but not on the branch for 6.3, e.g. see HandleNestTypesOnSourceJsonConverter, Line 48 of master vs. branch 6.3, see https://github.com/elastic/elasticsearch-net/pull/3357/commits/53d1c9688acd4d5efc51198eaeb8c40127753bd5

**Comment by russcam** [comment]
@baddonkey you are correct, looks like this commit was not backported for the 6.3.0 release 😞

**Comment by badpp** [comment]
Do you have any plans to fix this for 6.3.1 and when could you release it? We are totally blocked on this.

**Comment by russcam** [comment]
It's up on MyGet now in [NEST.JsonNetSerializer-6.4.0-ci20180914T071555](https://www.myget.org/feed/elasticsearch-net/package/nuget/NEST.JsonNetSerializer)

**Comment by badpp** [comment]
I switched temporarily our solution toNEST.JsonNetSerializer-6.4.0-ci20180914T071555, the unit tests regarding the serialization/deserialization are now succeeding. 

Unfortunately I've a good bunch of other unit tests now failing e.g. exceptions thrown on on bulk operations like `var response = await client.BulkAsync(descriptor).ConfigureAwait(false);` which throws directly an exception without the chance to check the response.IsValid flag.

`
            ----- Inner Stack Trace -----
   at Nest.RoutingResolver.ResolveId(Id id, IConnectionSettingsValues nestSettings)
   at Nest.RoutingResolver.Resolve(Type type, Object object)
   at Nest.RoutingResolver.Resolve[T](T object)
   at Nest.Routing.GetString(IConnectionSettingsValues nestSettings)
   at Nest.ElasticContractResolver.ShouldSerializeRouting(Object o, JsonProperty prop)
   at Nest.Json.Serialization.JsonSerializerInternalWriter.ShouldSerialize(JsonWriter writer, JsonProperty property, Object target)
   at Nest.Json.Serialization.JsonSerializerInternalWriter.CalculatePropertyValues(JsonWriter writer, Object value, JsonContainerContract contract, JsonProperty member, JsonProperty property, JsonContract& memberContract, Object& memberValue)
   at Nest.Json.Serialization.JsonSerializerInternalWriter.SerializeObject(JsonWriter writer, Object value, JsonObjectContract contract, JsonProperty member, JsonContainerContract collectionContract, JsonProperty containerProperty)
   at Nest.Json.Serialization.JsonSerializerInternalWriter.Serialize(JsonWriter jsonWriter, Object value, Type objectType)
   at Nest.Json.JsonSerializer.SerializeInternal(JsonWriter jsonWriter, Object value, Type objectType)
   at Nest.InternalSerializer.Serialize[T](T data, Stream writableStream, SerializationFormatting formatting)
   at Elasticsearch.Net.ElasticsearchSerializerExtensions.SerializeToBytes[T](IElasticsearchSerializer serializer, T data, SerializationFormatting formatting)
   at Nest.BulkRequestJsonConverter.WriteJson(JsonWriter writer, Object value, JsonSerializer serializer)
   at Nest.Json.Serialization.JsonSerializerInternalWriter.SerializeConvertable(JsonWriter writer, JsonConverter converter, Object value, JsonContract contract, JsonContainerContract collectionContract, JsonProperty containerProperty)
   at Nest.Json.Serialization.JsonSerializerInternalWriter.Serialize(JsonWriter jsonWriter, Object value, Type objectType)
   at Nest.Json.JsonSerializer.SerializeInternal(JsonWriter jsonWriter, Object value, Type objectType)
   at Nest.InternalSerializer.Serialize[T](T data, Stream writableStream, SerializationFormatting formatting)
   at Nest.InternalSerializer.SerializeAsync[T](T data, Stream stream, SerializationFormatting formatting, CancellationToken cancellationToken)
   at Elasticsearch.Net.SerializableData1.<WriteAsync>d__4.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Elasticsearch.Net.HttpWebRequestConnection.<RequestAsync>d__141.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Elasticsearch.Net.RequestPipeline.<CallElasticsearchAsync>d__741.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Elasticsearch.Net.Transport1.<RequestAsync>d__151.MoveNext()
Result Message:	
Elasticsearch.Net.UnexpectedElasticsearchClientException : Object reference not set to an instance of an object.
---- System.NullReferenceException : Object reference not set to an instance of an object.
`

Not sure whether this is a 6.4.x only problem or also in 6.3.0. This will need some time to investigate but I would be glad to do this on the 6.3.1 patch. Do you have any plans when to release it?

**Comment by Mpdreamz** [comment]
Thanks for letting us know @badpp 👍 looking to release `6.3.1` today or tomorrow.

**Comment by Mpdreamz** [comment]
Released `6.3.1` just now (might take sec to pop up on nuget.org). Thanks for re-raising this @badpp 👍
